File: cplusplus/199BinaryTreeRightSideView.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        vector<int> res;
        rightViewList(root, res, 0);
        
        return res;
    }
    
    void rightViewList(TreeNode *node, vector<int>& list, int level) {
        if (node == NULL) return;
        if (list.size() == level)
            list.push_back(node->val);
        rightViewList(node->right, list, level + 1);
        rightViewList(node->left, list, level + 1);
    }
};

File: java/199BinaryTreeRightSideView.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> res = new ArrayList<Integer>();
        if (root == null)
            return res;
        
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        
        while (!queue.isEmpty()) {
            int size = queue.size();
            for (int i = 0; i < size - 1; ++i) {
                TreeNode tn = queue.poll();
                if (tn.left != null)
                    queue.offer(tn.left);
                if (tn.right != null)
                    queue.offer(tn.right);
            }
            
            TreeNode tn = queue.poll();
            if (tn.left != null)
                queue.offer(tn.left);
            if (tn.right != null)
                queue.offer(tn.right);
            
            res.add(tn.val);
        }
        
        return res;
    }
}
